### PR TITLE
rest: add Prometheus storage adapter

### PR DESCRIPTION
**Pull request by sileht**

### CODE REVIEW

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 4751beabdae288fd1e2dc4c77eec8da11f3cd9b5

**Comment by sileht** [comment]
Pull-request updated, HEAD is now fef6c6da3a540c1302140d0336390a992d882c7e

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 43f3e3939d99a731ef788c4c6fb714dba93184be

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 826d910a8d7ffea9da786fc2cd37af43894b53cb

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 93fed111323dfc047e24a4cae1a21329ba75841c

**Comment by sileht** [comment]
Pull-request updated, HEAD is now a917d5290838e38ede8292ea17925a446307bbfc

**Comment by sileht** [comment]
Pull-request updated, HEAD is now e421f244f895f45c5493ae346a6af1b6a43363a6

**Comment by sileht** [comment]
Pull-request updated, HEAD is now d59224d2e7671c0165e9b10f26391963c19813ea

File: gnocchi/rest/api.py [hunk 1]
```diff
@@ -1883,6 +1890,62 @@ class BatchController(object):
...
+                need_update = True
+            else:
+                need_update = False
+        else:
+            need_update = True
+
+        if need_update:
+            if not r:
+                r = pecan.request.indexer.get_resource('generic', rid,
+                                                       with_metrics=True)
```
Comment (jd) [inline gnocchi/rest/api.py]: I don't want to kill the mood, but looks like it can also have been deleted at this stage and then raise an exception.

**Comment by sileht** [comment]
Pull-request updated, HEAD is now a5a1d951b3d9c34baacdc1601ea1b41c4f6640a2

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 2cb9792d73fbb364c344e964cf81c606e4c2edaf

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 2cb9792d73fbb364c344e964cf81c606e4c2edaf

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 5d4eacbc9e4558b3e53c0de07e8c3cf164063147

File: gnocchi/rest/api.py [hunk 2]
```diff
@@ -1892,6 +1900,67 @@ class BatchController(object):
...
+            ).metrics
+
+    @pecan.expose()
+    def post(self):
+        buf = snappy.uncompress(pecan.request.body)
+        f = remote_pb2.WriteRequest()
+        f.ParseFromString(buf)
+        measures = {}
+        for ts in f.timeseries:
+            attrs = dict((l.name, l.value) for l in ts.labels)
```
Comment (sileht) [inline gnocchi/rest/api.py]: I have reread the prometheus documentation and I have the feeling that Prometheus does uniquely identify metrics

I think putting all of them in "prometheus" resource will not help to query them.

Comment (sileht) [reply gnocchi/rest/api.py]: I have got how to identify a resource in prometheus. I will update the PR soon.

File: doc/source/prometheus.rst [hunk 3]
```diff
@@ -0,0 +1,12 @@
+====================
+ Prometheus support
+====================
+
+`Prometheus`_ can use Gnocchi to store its data through
+`Remote Write Adapter`_. Example of Prometheus configuration::
+
+  remote_write:
+  - url: "http://localhost:8041/v1/prometheus/write"
+
```
[the remark this answers is not in the document]
Comment (sileht) [reply doc/source/prometheus.rst]: I have added some doc about this.

**Comment by sileht** [comment]
Pull-request updated, HEAD is now ef75ec6a1c82c63782aa92749159f2ff544df93b

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 54b46067bb9fedf4c25725bf954676560b1795b7

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 42193828de0a608d0c0b2ddf407ec10a01aa5174

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 8c915e8a0c9ff729ff79a0047a472d341dce9cfe

File: doc/source/prometheus.rst [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+`Prometheus`_ can use Gnocchi to store its data through
+`Remote Write Adapter`_.
+
+Example of Prometheus configuration::
+
+  remote_write:
+  - url: "http://localhost:8041/v1/prometheus/write"
+
+
+The `/v1/prometheus/write` handles the `WriteRequest` protobuf message
```
Comment (jd) [inline doc/source/prometheus.rst]: The `/v1/prometheus/write` endpoint handles the `WriteRequest` protobuf message

File: doc/source/prometheus.rst [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+Gnocchi uses these attributes to identify the orignal resource polled by
+Prometheus.
+
+For each job/instance pair, we have a corresponding resource in Gnocchi.
+Prometheus resources are stored within the `prometheus` resource-type.
+The `original_resource_id` field is built like `<job>@<instance>`.
+A `prometheus` resource-type have to extra attributes `job` and `instance`.
+
+Then the `name` of the `Prometheus`_ metric is used as Gnocchi metric name.
+
```
Comment (jd) [inline doc/source/prometheus.rst]: Ok I think I get what you mean in this page. Here's a clearer version:


Gnocchi maps Prometheus metrics to its data model.

For each metric sent by Prometheus, Gnocchi maintains a corresponding resource based on each the `job` and `instance` pair. This resource is created with the `prometheus` resource type and contains two attributes, `job` and `instance`. The metrics sent by Prometheus with this pair are attached to that resource and filled with the provided measures.

File: gnocchi/rest/api.py [hunk 6]
```diff
@@ -1890,6 +1899,102 @@ class BatchController(object):
...
+        for (job, instance), measures in measures_by_rid.items():
+            original_rid = '%s@%s' % (job, instance)
+            rid = ResourceUUID(original_rid, creator=creator)
+            metric_names = list(measures.keys())
+            metrics = self.get_or_create_resource_and_metrics(
+                creator, rid, original_rid, job, instance, metric_names)
+
+            measures_to_batch.update(
+                dict((metric, measures[metric.name]) for metric in
+                     metrics if metric.name in measures))
```
Comment (jd) [inline gnocchi/rest/api.py]:
```
                measures_to_batch.update(
                    dict((metrics[metric_name], m)
                         for metric_name, m
                         in six.iteritems(measures)))
```
No?

File: doc/source/prometheus.rst [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+ Prometheus support
+====================
+
+`Prometheus`_ can use Gnocchi to store its data through
+`Remote Write Adapter`_.
+
+Example of Prometheus configuration::
+
+  remote_write:
+  - url: "http://localhost:8041/v1/prometheus/write"
```
Comment (jd) [inline doc/source/prometheus.rst]: This does not seem to handle auth and it's likely required.

File: gnocchi/rest/api.py [hunk 8]
```diff
@@ -1890,6 +1899,102 @@ class BatchController(object):
...
+                                         'value': s.value}
+                                        for s in ts.samples]))
+
+        creator = pecan.request.auth_helper.get_current_user(pecan.request)
+
+        measures_to_batch = {}
+        for (job, instance), measures in measures_by_rid.items():
+            original_rid = '%s@%s' % (job, instance)
+            rid = ResourceUUID(original_rid, creator=creator)
+            metric_names = list(measures.keys())
```
Comment (jd) [inline gnocchi/rest/api.py]: It's possible Prometheus supports `/` in metric name and Gnocchi does not. You need to do a replace I guess and document it.

File: gnocchi/rest/api.py [hunk 9]
```diff
@@ -1890,6 +1899,102 @@ class BatchController(object):
...
+                     "max_length": 512,
+                     "required": True},
+        "job": {"type": "string",
+                "min_length": 1,
+                "max_length": 512,
+                "required": True}
+    }
+
+    @classmethod
+    @tenacity.retry(retry=tenacity.retry_if_exception_type(
```
Comment (jd) [inline gnocchi/rest/api.py]: You don't want to retry forever. We need a time limit.

I think we have a time limit for refresh already. We can use that.

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 89f0d3a85c49f02a71ee943efd06ffb7e3f38f13

**Comment by sileht** [comment]
Pull-request updated, HEAD is now eafb274f39ed5609390a666a5b27564cc3cf8c0b

File: gnocchi/rest/api.py [hunk 10]
```diff
@@ -1890,6 +1899,102 @@ class BatchController(object):
...
+                                           original_resource_id,
+                                           job, instance, metric_names):
+        try:
+            r = pecan.request.indexer.get_resource('prometheus', rid,
+                                                   with_metrics=True)
+        except indexer.NoSuchResourceType:
+            schema = pecan.request.indexer.get_resource_type_schema()
+            rt = schema.resource_type_from_dict(
+                'prometheus', cls.PROMETHEUS_RESOURCE_TYPE, 'creating')
+            pecan.request.indexer.create_resource_type(rt)
```
Comment (jd) [inline gnocchi/rest/api.py]: do you miss ResourceTypeAlreadyExists here?

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 14446bff4786eb94c51a0d3ba3df9a99a26c50ba

File: gnocchi/rest/api.py (hunk 8 above)
Comment (sileht) [reply gnocchi/rest/api.py]: I missed that, fixed.

File: doc/source/prometheus.rst (hunk 7 above)
Comment (sileht) [reply doc/source/prometheus.rst]: fixed

File: gnocchi/rest/api.py (hunk 8 above)
Comment (sileht) [reply gnocchi/rest/api.py]: I missed that, fixed.

File: doc/source/prometheus.rst (hunk 4 above)
Comment (sileht) [reply doc/source/prometheus.rst]: done

File: doc/source/prometheus.rst (hunk 5 above)
Comment (sileht) [reply doc/source/prometheus.rst]: done

File: gnocchi/rest/api.py (hunk 9 above)
Comment (sileht) [reply gnocchi/rest/api.py]: If that retry more that once, I prefer not to be guy that do the sql maintenance operation in this case :), Fixed

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 940d7e5897747703719304ded8a12cdc6b8db4b2

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 8e0ffd55a045aab69f649533b121c6f11c99c460

File: doc/source/prometheus.rst [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+      password: "whatever"
+
+
+The `/v1/prometheus/write` endpoint handles the `WriteRequest` protobuf
+message.
+
+Gnocchi maps Prometheus metrics to its data model.
+
+For each metric sent by Prometheus, Gnocchi maintains a corresponding resource
+based on each the `job` and `instance` pair. This resource is created with the
```
Comment (jd) [inline doc/source/prometheus.rst]: s/the//

File: gnocchi/rest/api.py [hunk 12]
```diff
@@ -1890,6 +1899,106 @@ class BatchController(object):
...
+        wait=tenacity.wait_exponential(multiplier=0.5, max=60),
+        retry=tenacity.retry_if_exception_type(
+            (indexer.NoSuchResource, indexer.ResourceAlreadyExists,
+             indexer.NamedMetricAlreadyExists)))
+    def get_or_create_resource_and_metrics(cls, creator, rid,
+                                           original_resource_id,
+                                           job, instance, metric_names):
+        try:
+            r = pecan.request.indexer.get_resource('prometheus', rid,
+                                                   with_metrics=True)
```
Comment (jd) [inline gnocchi/rest/api.py]: If `rid` already exists with type != `prometheus` then r will be None, but `create_resource` on L1956 will raise `indexer.ResourceAlreadyExists`, which will make this retry, and loop forever.

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 33d0e02112130660b58802a434f5920c1d414c34

File: gnocchi/rest/api.py (hunk 12 above)
Comment (sileht) [reply gnocchi/rest/api.py]: I have fixed it, but honestly I will be a uuid collision, so...

File: gnocchi/rest/api.py [hunk 13]
```diff
@@ -1890,6 +1899,114 @@ class BatchController(object):
...
+            try:
+                return pecan.request.indexer.create_resource(
+                    'prometheus', rid, creator,
+                    original_resource_id=original_resource_id,
+                    job=job,
+                    instance=instance,
+                    metrics=metrics
+                ).metrics
+            except indexer.ResourceAlreadyExists:
+                # NOTE(sileht): ensure the rid is not registered whitin another
```
Comment (jd) [inline gnocchi/rest/api.py]: within or with

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 65aafd7bf6f80849c782dd8b5f576bd3f2b10a9f

**Comment by sileht** [comment]
Pull-request updated, HEAD is now aed84851af131bd9b9aa4b56979501792082c825

File: gnocchi/rest/api.py [hunk 14]
```diff
@@ -1890,6 +1899,118 @@ class BatchController(object):
     resources = ResourcesBatchController()
 
 
+class PrometheusWriteController(rest.RestController):
```
Comment (jd) [inline gnocchi/rest/api.py]: Not a big issue but is there any reason you did not put that into `gnocchi.rest.prometheus` to not make that file too big?

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 6acd894012e3624b24cd2fc4d235ac4678abec7b

**Comment by sileht** [comment]
Pull-request updated, HEAD is now e73f4cb5a12d7adb0f365465d3cfbdc961f5c91e

File: gnocchi/rest/api.py [hunk 15]
```diff
@@ -1890,6 +1899,118 @@ class BatchController(object):
...
+                    metrics=metrics,
+                    append_metrics=True,
+                    create_revision=False
+                ).metrics
+            else:
+                return r.metrics
+        else:
+            metrics = MetricsSchema(dict((m, {}) for m in metric_names))
+            try:
+                return pecan.request.indexer.create_resource(
```
Comment (jd) [inline gnocchi/rest/api.py]: This also should check for ACL IMHO

File: gnocchi/rest/api.py [hunk 16]
```diff
@@ -1890,6 +1899,118 @@ class BatchController(object):
...
+                raise tenacity.TryAgain
+
+        if r:
+            exists_metric_names = [m.name for m in r.metrics]
+            metrics = MetricsSchema(dict(
+                (m, {}) for m in metric_names
+                if m not in exists_metric_names
+            ))
+            if metrics:
+                return pecan.request.indexer.update_resource(
```
Comment (jd) [inline gnocchi/rest/api.py]: This should check for ACL

File: gnocchi/rest/api.py [hunk 17]
```diff
@@ -1890,6 +1899,118 @@ class BatchController(object):
...
+                                           original_resource_id,
+                                           job, instance, metric_names):
+        try:
+            r = pecan.request.indexer.get_resource('prometheus', rid,
+                                                   with_metrics=True)
+        except indexer.NoSuchResourceType:
+            schema = pecan.request.indexer.get_resource_type_schema()
+            rt = schema.resource_type_from_dict(
+                'prometheus', cls.PROMETHEUS_RESOURCE_TYPE, 'creating')
+            pecan.request.indexer.create_resource_type(rt)
```
Comment (jd) [inline gnocchi/rest/api.py]: ACL check?

**Comment by sileht** [comment]
Pull-request updated, HEAD is now d0324a8498324167c84a01f3f18fcecadd7db4c1

**Comment by sileht** [comment]
Pull-request updated, HEAD is now 314ed651ac1f52e828cc4c964aa2694b75612fe7

File: gnocchi/rest/api.py [hunk 18]
```diff
@@ -1890,6 +1899,136 @@ class BatchController(object):
...
+            enforce("create resource type", {
+                'name': 'prometheus',
+                'state': 'creating',
+                'attributes': cls.PROMETHEUS_RESOURCE_TYPE
+            })
+
+            schema = pecan.request.indexer.get_resource_type_schema()
+            rt = schema.resource_type_from_dict(
+                'prometheus', cls.PROMETHEUS_RESOURCE_TYPE, 'creating')
+            pecan.request.indexer.create_resource_type(rt)
```
Comment (chungg) [inline gnocchi/rest/api.py]: can this throw an exception too? if  say two things try to create same resource_type at same time?

Comment (jd) [reply gnocchi/rest/api.py]: IIUC this can raise `indexer.ResourceTypeAlreadyExists` which is gonna make the code retry (see the tenacity decorator).

**Pull request merged**